data/registry_model/quota.py

```diff
@@ -0,0 +1,353 @@
+            )
+            .get()
+        )
+        return True
+    except ManifestBlob.DoesNotExist:
+        return False
+
+
+def blob_exists_in_repository(repository_id: int, manifest_id: int, blob_id: int):
+    try:
+        (
+            ManifestBlob.select(ManifestBlob.id)
+            .where(
+                ManifestBlob.repository == repository_id,
+                ManifestBlob.blob == blob_id,
+                ManifestBlob.manifest != manifest_id,
+            )
+            .get()
+        )
+        return True
+    except ManifestBlob.DoesNotExist:
+        return False
+
+
+def write_namespace_total(
+    namespace_id: int, manifest_id: int, namespace_total: int, operation: str
+):
+    namespace_size = get_namespace_size(namespace_id)
+    namespace_size_exists = namespace_size is not None
+
+    # If backfill hasn't ran yet for this namespace don't do anything
+    if namespace_size_exists and (
+        namespace_size.backfill_start_ms is None
+        or namespace_size.backfill_start_ms > get_epoch_timestamp_ms()
+    ):
+        return
+
+    # If the namespacesize entry doesn't exist and this is the only manifest in the namespace
+    # we can assume this is the first push to the namespace and there is no blobs to be
+    # backfilled, so let the entry be created. Otherwise it still needs to be handled by the
+    # backfill worker so let's exit
+    params = {}
+    if (
+        operation == "add"
+        and not namespace_size_exists
+        and only_manifest_in_namespace(namespace_id, manifest_id)
+    ):
+        params["backfill_start_ms"] = 0
+        params["backfill_complete"] = True
+    elif operation == "add" and not namespace_size_exists:
+        return
+
+    increment_namespacesize(namespace_id, namespace_total, operation, namespace_size_exists, params)
+
+
+def write_repository_total(
+    repository_id: int, manifest_id: int, repository_total: int, operation: str
+):
+    repository_size = get_repository_size(repository_id)
+    repository_size_exists = repository_size is not None
+
+    # If backfill hasn't ran yet for this repository don't do anything
+    if repository_size_exists and (
+        repository_size.backfill_start_ms is None
+        or repository_size.backfill_start_ms > get_epoch_timestamp_ms()
+    ):
+        return
+
+    # If the repositorysize entry doesn't exist and this is the only manifest in the repository
+    # we can assume this is the first push to the repository and there is no blobs to be
+    # backfilled, so let the entry be created. Otherwise it still needs to be handled by the
+    # backfill worker so let's exit
+    params = {}
+    if (
+        operation == "add"
+        and not repository_size_exists
+        and only_manifest_in_repository(repository_id, manifest_id)
+    ):
+        params["backfill_start_ms"] = 0
+        params["backfill_complete"] = True
+    elif operation == "add" and not repository_size_exists:
+        return
```

Comment:
Ditto. I would just let the caller figure out if this needs to be called or not.

Reply:
Yeah that's the approach [here.](https://github.com/quay/quay/pull/1751/files#diff-4ecb6ccf56f6cd4aaec2d17fb7224c8264834d10fb8db833830a8bae5ff90dd8R26-R32)